holoviews/plotting/bokeh/element.py

```diff
@@ -492,28 +507,67 @@ def _update_plot(self, key, plot, element=None):
 
     def _update_ranges(self, element, ranges):
         plot = self.handles['plot']
-        l, b, r, t = self.get_extents(element, ranges)
-
-        if self.invert_axes:
-            l, b, r, t = b, l, t, r
-        if l == r:
-            offset = abs(l*0.1 if l else 0.5)
-            l -= offset
-            r += offset
-        if b == t:
-            offset = abs(b*0.1 if b else 0.5)
-            b -= offset
-            t += offset
-
-        # Ensure that it never sets a NaN value
-        if isinstance(plot.x_range, Range1d):
+        x_range = self.handles['x_range']
+        y_range = self.handles['y_range']
+
+        if any(isinstance(r, Range1d) for r in [x_range, y_range]):
+            l, b, r, t = self.get_extents(element, ranges)
+            if self.invert_axes:
+                l, b, r, t = b, l, t, r
+
+        if any(isinstance(r, FactorRange) for r in [x_range, y_range]):
+            xfactors, yfactors = self._get_factors(element)
+
+        if isinstance(x_range, Range1d):
+            if l == r:
+                offset = abs(l*0.1 if l else 0.5)
+                l -= offset
+                r += offset
             if isinstance(l, np.datetime64) or np.isfinite(l): plot.x_range.start = l
-            if isinstance(l, np.datetime64) or np.isfinite(r): plot.x_range.end   = r
+            if isinstance(r, np.datetime64) or np.isfinite(r): plot.x_range.end   = r
+        elif isinstance(x_range, FactorRange):
+            x_range.factors = list(xfactors)
+
         if isinstance(plot.y_range, Range1d):
+            if b == t:
+                offset = abs(b*0.1 if b else 0.5)
+                b -= offset
+                t += offset
             if isinstance(l, np.datetime64) or np.isfinite(b): plot.y_range.start = b
             if isinstance(l, np.datetime64) or np.isfinite(t): plot.y_range.end   = t
+        elif isinstance(y_range, FactorRange):
+            y_range.factors = list(yfactors)
 
 
+    def _clean_data(self, data, cols, dims):
+        """
+        Cleans the data before instantiating the datasource to handle
+        categorical axes correctly.
+        """
```

Comment:
I think calling it ``_categorize_data`` and explicitly mentioning the string conversion/pretty printing in the docstring would be clearer.